Math/NumberTheory/EisensteinIntegers.hs

```diff
@@ -0,0 +1,68 @@
+-- |
+-- Module:      Math.NumberTheory.EisensteinIntegers
+-- Licence:     MIT
+-- Stability:   Provisional
+-- Portability: Non-portable (GHC extensions)
+--
+-- This module exports functions for manipulating Eisenstein integers, including
+-- computing their prime factorisations.
+--
+
+{-# LANGUAGE DeriveGeneric #-}
+
+module Math.NumberTheory.EisensteinIntegers
+  ( EisensteinInteger(..)
+  ,  ω
+  , divE
+  ) where
+
+import GHC.Generics (Generic)
+
+infix 6 :+
+
+-- |An Eisenstein integer is a + bω, where a and b are both integers.
+data EisensteinInteger = (:+) { real :: !Integer, imag :: !Integer }
+    deriving (Eq, Ord, Generic)
+
+-- | The imaginary unit for Eisenstein integers, where
+--
+-- > ω == (-1/2) + ((sqrt 3)/2)ι == exp(2*pi*ι/3)
+-- and ι is the usual imaginary unit with ι² == -1.
+ω :: EisensteinInteger
+ω = 0 :+ 1
+
+instance Show EisensteinInteger where
+    show (a :+ b)
+        | b == 0     = show a
+        | a == 0     = s ++ b'
+        | otherwise  = show a ++ op ++ b'
+        where
+            b' = if abs b == 1 then "ω" else show (abs b) ++ "*ω"
+            op = if b > 0 then "+" else "-"
+            s  = if b > 0 then "" else "-"
+
+instance Num EisensteinInteger where
+    (+) (a :+ b) (c :+ d) = (a + c) :+ (b + d)
+    (*) (a :+ b) (c :+ d) = (a * c - b * d) :+ (b * c + a * d - b * d)
+    -- An Eisenstein integer @a :+ b@, with @a, b@ integers, can we written as
+    -- @(2*a - b) / 2 + ((b * sqrt 3) * ι) / 2@, but this number is in the
+    -- same quadrant as @(2*a - b) / 2 + (b * ι) / 2@, and this one in the
+    -- same as @(2*a - b) + b * ι@. Divisions or floating points are not
+    -- necessary.
+    abs z@(x :+ y) = abs' (2*x - y) x
```

Comment:
@Bodigrim I just realized this is wrong. First of all, what one take the absolute value of an Eisenstein integer, `abs n` to be? Is it an E. integer `n'` with the same magnitude such that `(signum n) * (abs n) == n'`?

With Gaussian integers, there are four units, one for each quadrant of the complex plane, so `signum` always returns one of those.
However, the commutative ring of Eisenstein integers has 6 units, `{1, -1, ω, -ω, ω², -ω²}`, one cannot assign each to a regular Cartesian quadrant. Do we divide the complex plane into 6 parts when dealing with Eisenstein integers?

Reply:
Exactly, there are 6 sextants, 60 degrees each. `abs` should put its argument into the first one.